components/common/src/test/java/com/hotels/styx/config/validator/DocumentFormatTest.java

```diff
@@ -0,0 +1,526 @@
+        newDocument()
+                .subSchema("PluginsList", schema(field("all", string())))
+                .rootSchema(schema(
+                        field("parent", object(
+                                field("httpPipeline", object("notExists"))
+                        )),
+                        field("plugins", object("PluginsList"))
+                ))
+                .build();
+    }
+
+    @Test(expectedExceptions = InvalidSchemaException.class,
+            expectedExceptionsMessageRegExp = "No schema configured for lazy object reference 'non-existing-HandlerConfig'")
+    public void documentBuilderEnsuresNamedSchemaReferencesAreValidFromSubSchemas() throws Exception {
+        newDocument()
+                .subSchema("HttpPipeline", schema(
+                        field("handlers", object("non-existing-HandlerConfig"))
+                ))
+                .subSchema("PluginsList", schema(field("all", string())
+                ))
+                .rootSchema(schema(
+                        field("plugins", object("PluginsList")),
+                        field("httpPipeline", object("HttpPipeline"))
+                ))
+                .build();
+    }
+
+    @Test
+    public void validatesDiscriminatedUnions() throws Exception {
+
+        DocumentFormat validator = newDocument()
+                .subSchema("ProxyTo", schema(
+                        field("id", string()),
+                        field("destination", string())
+                ))
+                .subSchema("Redirection", schema(
+                        field("status", integer()),
+                        field("location", string())
+                ))
+                .rootSchema(schema(
+                        field("httpPipeline", object(
+                                field("type", string()),
+                                field("config", union("type"))
+                        ))
+                ))
+                .build();
+
+
+        boolean outcome1 = validator.validateObject(
+                YAML_MAPPER.readTree(
+                        "httpPipeline: \n"
+                                + "  type: 'ProxyTo'\n"
+                                + "  config:\n"
+                                + "    id: 'local-01'\n"
+                                + "    destination: 'localhost:8080'\n"
+                ));
+        assertThat(outcome1, is(true));
+
+        boolean outcome2 = validator.validateObject(
+                YAML_MAPPER.readTree(
+                        "httpPipeline: \n"
+                                + "  type: 'Redirection'\n"
+                                + "  config:\n"
+                                + "    status: 301\n"
+                                + "    location: /new/location\n"
+                ));
+        assertThat(outcome2, is(true));
+
+    }
+
+    @Test
+    public void ignoresSubobjects() throws Exception {
```

Comment:
Should be something like `ignoresSubobjectsWhenOpaqueDirectiveIsUsed`.
Also, it might be best to replace this neologism "subobject" with something like "child object", which is more familiar terminology.

Reply:
It is going to say "ignoresOpaqueSubobjectValidation".